pkg/assets/extensions.go

```diff
@@ -0,0 +1,163 @@
+package assets
+
+import (
+	"bytes"
+	"crypto/md5"
+	"encoding/hex"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"k8s.io/kubernetes/pkg/util"
+)
+
+// ExtensionJSHandler concatenates and serves extension JavaScript files as one HTTP response.
+func ExtensionJSHandler(files []string) (http.Handler, error) {
+	bytes, err := concatAll(files, ";\n")
+	if err != nil {
+		return nil, err
+	}
+	etag := calculateETag(bytes)
+
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		serve(w, r, bytes, "text/javascript", etag)
+	}), nil
+}
+
+// ExtensionCSSHandler concatenates and serves extension stylesheets as one HTTP response.
+func ExtensionCSSHandler(files []string) (http.Handler, error) {
+	bytes, err := concatAll(files, "\n")
+	if err != nil {
+		return nil, err
+	}
+	etag := calculateETag(bytes)
+
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		serve(w, r, bytes, "text/css", etag)
+	}), nil
+}
+
+func concatAll(files []string, separator string) ([]byte, error) {
+	var buffer bytes.Buffer
+	for _, file := range files {
+		f, err := os.Open(file)
+		if err != nil {
+			return nil, err
+		}
+		defer f.Close()
+
+		_, err = buffer.ReadFrom(f)
+		if err != nil {
+			return nil, err
+		}
+
+		_, err = buffer.WriteString(separator)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return buffer.Bytes(), nil
+}
+
+func calculateETag(bytes []byte) string {
+	// Calculate an ETag from the MD5 hash
+	hasher := md5.New()
+	hasher.Write(bytes)
+	sum := hasher.Sum(nil)
+
+	// ETag values are quoted
+	return fmt.Sprintf("\"%s\"", hex.EncodeToString(sum))
```

Comment:
it is weak because it doesn't include the entire entity, including headers.

the presence of the Vary header isn't sufficient... the ETag has to be different for different Vary headers, see https://bz.apache.org/bugzilla/show_bug.cgi?id=46538 for an example explanation